### PR TITLE
TextPanel: Support code formats

### PR DESCRIPTION
**Pull request by ryantxu**
This adds support to show the monoco code editor in the text panel when "code" mode is selected:

image

TODO:
- [x] devenv dashboard.  See: http://localhost:3000/d/WZ7AhQiVz/text-options
- [x] update docs

### CODE REVIEW

File: devenv/dev-dashboards/panel-text/text-options.json
Comment (ryantxu) [inline devenv/dev-dashboards/panel-text/text-options.json:318]: @sdboyer -- FYI newly added dashboard that passed CI on the first try 🎉

Comment (sdboyer) [reply devenv/dev-dashboards/panel-text/text-options.json:318]: tremendous!

i look forward to when this is an utterly forgettable event :)

File: docs/sources/visualizations/text-panel.md
Comment (chri2547) [inline docs/sources/visualizations/text-panel.md:32]:
```suggestion
This setting renders the content as [sanitized](https://github.com/grafana/grafana/blob/code-in-text-panel/packages/grafana-data/src/text/sanitize.ts) HTML. If you require more direct control over the output, you can set the 
[disable_sanitize_html]({{< relref "../setup-grafana/configure-grafana/#disable_sanitize_html" >}}) flag which enables you to directly enter HTML.
```

File: docs/sources/visualizations/text-panel.md
Comment (chri2547) [inline docs/sources/visualizations/text-panel.md:26]:
```suggestion
This option formats the content as [markdown](https://en.wikipedia.org/wiki/Markdown).
```

File: docs/sources/visualizations/text-panel.md
Comment (chri2547) [inline docs/sources/visualizations/text-panel.md:36]:
```suggestion
This option renders content inside a read-only code editor. The selected language will provide proper syntax highlighting
```

Comment (chri2547) [inline docs/sources/visualizations/text-panel.md:36]: I'm unsure of what you are trying to say in this sentence "The selected language will provide proper syntax highlighting you want to use markdown or HTML to style your text, then enter content in the box below."

Could you rephrase that and I can take another look?

**Review by chri2547 — CHANGES_REQUESTED** [review]
A few suggestions and a question. Thank you!

File: docs/sources/visualizations/text-panel.md
Comment (ryantxu) [reply docs/sources/visualizations/text-panel.md]: 🤦 -- that was a garble with the previous language.  Updated and lets see if it makes more sense :)

**Review by torkelo — COMMENTED** [review]
Feels strange (and expensive) to show code as a read-only text area using monaco. If it's just for syntax highlighting maybe using Prism.highlight is an option? That should be much faster than doing syntax highlighting via Monaco. 

https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/prometheus/querybuilder/shared/RawQuery.tsx#L18 

If Monaco is required for more reasons can we style it as not an input/text area ?

**Review by chri2547 — COMMENTED** [review]
The docs LGTM! Please go ahead and merge after Engr review and approval.

**Comment by ryantxu** [comment]
I get that using monaco for syntax highlighting feels like a big hammer...  it feels bigger/riskier to add a new tool (prism/ace/etc) to do the same thing, but less supported and not already used throughout our code.

My biggest question is if this should be its own panel, or a different one -- in particular the goals from https://github.com/grafana/grafana/pull/41095 and https://grafana.com/grafana/plugins/ryantxu-ajax-panel/ could be supported seemlessly within the text panel.  But if we want/need to expand what context+tools the panel has, then it may make sense to create a new panel.  Though I am not sure what we would call it.

**Comment by ryantxu** [comment]
Also re monoco vs simple syntax highlighting -- for JSON, it also does the nice code folding (expand/collapse) very efficiently.  I have not found anything else that supports a large body and does not crash/hang/freeze.

**Comment by torkelo** [comment]
Made it so that the code editor uses primary background so it blends more into a being panel rather than a panel with a huge text area (which looks bad on a dashboard as the canvas color is the same as the input background). 

Screenshot from 2022-08-23 08-48-56

> My biggest question is if this should be its own panel, or a different one -- in particular the goals from https://github.com/grafana/grafana/pull/41095 and https://grafana.com/grafana/plugins/ryantxu-ajax-panel/ could be supported seemlessly within the text panel. But if we want/need to expand what context+tools the panel has, then it may make sense to create a new panel. Though I am not sure what we would call it.'

I think showing code syntax highlighted is ok for the text panel, but maybe doing "templated" text rendering based on data should be a different panel (that requires queries).

**Review by torkelo — APPROVED** [review]

**Review by chri2547 — APPROVED** [review]

**Pull request merged**
